File: 백준/Silver/3085. 사탕 게임/사탕 게임.cc
#include <iostream>
#include <string>
using namespace std;

int N, ans;
char board[50][50];
void swap(char &a, char&b) {
    char tmp = a; a = b; b = tmp;
}
void checkRow(int row) {
    int cnt = 1;
    for(int i=1; i<N; i++) {
        if(board[row][i] == board[row][i-1]) {
            cnt++;
            if(cnt > ans) ans = cnt;
        } else cnt = 1;
    }
}
void checkCol(int col) {
    int cnt = 1;
    for(int i=1; i<N; i++) {
        if(board[i][col] == board[i-1][col]) {
            cnt++;
            if(cnt > ans) ans = cnt;
        } else cnt = 1;
    }
}
int main() {
    ios::sync_with_stdio(0); cin.tie(0);
    cin >> N;
    for(int i=0; i<N; i++) cin >> board[i];
    for(int i=0; i<N; i++) {
        for(int j=0; j<N; j++) {
            if(i < N-1) {
                swap(board[i][j], board[i+1][j]);
                checkCol(j);
                checkRow(i);
                checkRow(i+1);
                swap(board[i][j], board[i+1][j]);
            }
            if(j < N-1) {
                swap(board[i][j], board[i][j+1]);
                checkRow(i);
                checkCol(j);
                checkCol(j+1);
                swap(board[i][j], board[i][j+1]);
            }
        }
    }
    cout << ans;
}

File: 백준/Silver/3085. 사탕 게임/사탕 게임.py
import sys
In = sys.stdin.readline
N = int(In())
graph = [list(In().rstrip()) for _ in range(N)]

def checkRow(row):
    result = cnt = 1
    for col in range(1, N):
        cnt = cnt+1 if graph[row][col-1] == graph[row][col] else 1
        result = max(result, cnt)
    return result

def checkCol(col):
    result = cnt = 1
    for row in range(1, N):
        cnt = cnt+1 if graph[row-1][col] == graph[row][col] else 1
        result = max(result, cnt)
    return result

def swap(row1, col1, row2, col2):
    graph[row1][col1], graph[row2][col2] = graph[row2][col2], graph[row1][col1]

ans = 1
for i in range(N):
    ans = max(ans, checkCol(i), checkRow(i))

if(ans < N):
    for i in range(N):
        for j in range(N):
            if i<N-1 and graph[i][j] != graph[i+1][j]:
                swap(i,j, i+1,j)
                ans = max(ans, checkRow(i), checkRow(i+1), checkCol(j))
                swap(i,j, i+1,j)
            if j<N-1 and graph[i][j] != graph[i][j+1]:
                swap(i,j, i,j+1)
                ans = max(ans, checkCol(j), checkCol(j+1), checkRow(i))
                swap(i,j, i,j+1)
print(ans)